src/benchmark.ts

```diff
@@ -1,19 +1,139 @@
-export abstract class Benchmark {
-  startTime: number;
+import { errorToObject } from "./util";
 
-  constructor(private name: string) {}
+const DEFAULT_CONFIG: any = {
+  timeout: 10000
+};
 
-  async start() {
-    console.log(`Starting benchmark: ${this.name}`);
-    this.startTime = Date.now();
-  }
-
-  abstract async run(): Promise<void>;
+export abstract class Benchmark<
+  C extends BenchmarkConfig,
+  R extends BenchmarkReport
+> {
+  protected aborted = false;
+  protected config: C;
+  protected report: R = { execution: {}, result: {} } as R;
 
-  async end() {
-    console.log(
-      `End benchmark: ${this.name}. Total time: ${Date.now() -
-        this.startTime} ms`
+  constructor(config: C, defaultConfig: any) {
+    this.config = Object.assign(
+      {},
+      Object.assign(DEFAULT_CONFIG, defaultConfig),
+      config
     );
+    if (!this.config.name) {
+      throw new Error("config.name is required");
+    }
+    if (!this.config.type) {
+      throw new Error("config.type is required");
+    }
+    if (this.config.timeout <= 0) {
+      throw new Error("config.timeout must be a positive number");
+    }
+    this.report.config = this.config;
   }
+
+  public execute = (): Promise<R> => {
```

Comment:
I sometimes use arrow functions too much since I do not clearly understand when "this" ref is available or I need to mess with these `this.func = this.func.bind(this)` in constructor..